Mathlib/RingTheory/Polynomial/Hilbert.lean

```diff
@@ -0,0 +1,92 @@
+some `h : ℚ[X]` such that for any large enough `n : ℕ`, `h(n)` is equal to the coefficient of `Xⁿ`
+in the power series expansion of `p/(1 - X)ᵈ` (or `p * (PowerSeries.invOneSubPow ℤ d)`). This `h` is
+unique and is called the Hilbert polynomial with respect to `p` and `d` (`Polynomial.hilbert p d`).
+
+The above fact is used to construct the Hilbert polynomial of a graded module that satisfies certain
+conditions. Specifically:
+* Assume `A = ⊕ₙAₙ` is a Noetherian ring graded by `ℕ` such that `A` is generated by `a₁,...,aₛ` as
+  an `A₀`-algebra, where for each `i = 1,...,s`, `aᵢ` is a homogeneous element in `A` of degree
+  `dᵢ > 0`. Let `M = ⊕ₙMₙ` be a finitely generated `A`-module graded by `ℕ`. Then it is true that
+  each `Mₙ` is a finitely generated module over `A₀`.
+* Let `λ : C → ℤ` be an additive function, where `C` is the collection of all finitely generated
+  `A₀`-modules; in other words, given any short exact sequence `0 ⟶ N ⟶ O ⟶ P ⟶ 0` of finitely
+  generated modules over `A₀`, we have `λ(O) = λ(N) + λ(P)`. Then the Poincaré series of `M` in
+  terms of `λ` is the formal power series `P(M, λ, X) = Σₙλ(Mₙ)Xⁿ`. The Hilbert-Serre Theorem states
+  that `P(M, λ, X)` can be written as `p(X)/∏ᵢ₌₁,...,ₛ(1 - X ^ dᵢ)`, where `p(X)` is a polynomial
+  with coefficients in `ℤ`.
+* For the case when `d₁,...,dₛ = 1`, the Poincaré series of `M` with respect to `λ` can be expressed
+  as `p(X)/(1 - X)ˢ`. Hence the fact stated in the beginning guarantees an `h : ℚ[X]` such that for
+  any large enough `n : ℕ`, the coefficient of `Xⁿ` in `P(M, λ, X)` equals `h.eval (n : ℚ)`. This
+  `h` is called the Hilbert polynomial of `M` in terms of `λ`, which is what we eventually want to
+  formalise.
+-/
+
+open BigOperators Nat PowerSeries
+
+namespace Polynomial
+
+section greatestFactorOneSubNotDvd
+
+variable {R : Type*} [CommRing R] (p : R[X]) (hp : p ≠ 0) (d : ℕ)
+
+/--
+Given a polynomial `p`, the factor `f` of `p` such that the product of `f` and
+`(1 - X : R[X]) ^ p.rootMultiplicity 1` equals `p`. We define this here because if `p` is divisible
+by `1 - X`, then the expression `p/(1 - X)ᵈ` can be reduced. We want to construct the Hilbert
+polynomial based on the most reduced form of the fraction `p/(1 - X)ᵈ`. Later we will see that this
+method of construction makes it much easier to calculate the specific degree of the Hilbert
+polynomial.
+-/
+noncomputable def greatestFactorOneSubNotDvd : R[X] :=
+  ((- 1 : R[X]) ^ p.rootMultiplicity 1) *
+  (exists_eq_pow_rootMultiplicity_mul_and_not_dvd p hp 1).choose
+
+end greatestFactorOneSubNotDvd
+
+/--
+A polynomial which makes it easier to define the Hilbert polynomial. See also the theorem
+`Polynomial.preHilbert_eq_choose_sub_add`, which states that for any `d k n : ℕ` with `k ≤ n`,
+`(Polynomial.preHilbert d k).eval (n : ℚ) = (n - k + d).choose d`.
+-/
+noncomputable def preHilbert (d k : ℕ) : ℚ[X] :=
+  (d.factorial : ℚ)⁻¹ • (∏ i : Finset.range d, (X - (C (k : ℚ)) + (C (i : ℚ)) + 1))
```

Comment:
I wonder whether you can more API if you define the product as `(ascPochhammer S _).comp (X + _)` or `(ascPochhammer S _).taylor _` instead

Reply:
> I wonder whether you can more API if you define the product as `(ascPochhammer S _).comp (X + _)` or `(ascPochhammer S _).taylor _` instead

Thank you! I have changed the definition into `(d.factorial : ℚ)⁻¹ • ((ascPochhammer ℚ d).comp (X - (C (k : ℚ)) + 1))`, and consequently the proof of `preHilbert_eq_choose_sub_add` has become significantly shorter!